Fix spurious error from "docker load"

rcowsill opened the pull request:
**- What I did**
Fixed the spurious error "...is not a valid parent for..." when using `docker load`. Closes #41829

**- How I did it**
Added an `Equal` method to History, which normalizes UTC-equivalent times into UTC before comparison. This is used instead of `reflect.DeepEqual` in checkValidParent.

**- How to verify it**
With the following Dockerfile, on a Windows machine:
```
FROM mcr.microsoft.com/windows/nanoserver:1809
CMD ["cmd", "/C", "echo hello world"]
```
1. Run `docker build -t test-image .`
2. Run `docker save -o images.tar test-image mcr.microsoft.com/windows/nanoserver:1809`
3. Run `docker load -i images.tar`

With the patch this should work without giving an "is not a valid parent for" error message in step 3.

**- Description for the changelog**
Fix "is not a valid parent for" error from `docker load`

**- A picture of a cute animal (not mandatory but encouraged)**
🦊

tonistiigi commented on image/image.go:
```diff
@@ -218,6 +219,30 @@ func NewHistory(author, comment, createdBy string, isEmptyLayer bool) History {
+	h.normalize()
+	i.normalize()
+
+	return reflect.DeepEqual(h, i)
```
Should be cleaner to reset `Created` property and run `h.Created.Equal(i.Created)` for it specifically.

rcowsill replied:
I considered doing that, but it would mean that 14:00:00+01:00 would equal 16:00:00+03:00, for example.

Since the difference between those times is preserved across Unmarshal/Marshal (unlike "+00:00" vs "Z") I went for this approach.

thaJeztah replied:
(Thinking out loud); At a quick glance, I'd say that'd be correct 🤔; if we'd treat them as "different", that would mean that setting a different timezone on a server would be considered a different artefact even if the time was actually the same.

rcowsill replied:
For a valid parent the histories being compared were copied directly from parent to child. The only difference is that the child's copies took an extra round trip through UnmarshalJSON/MarshalJSON. In that case I don't think changing the server TZ can change the result of the comparison.

That said, I'll take another look at Time.Unmarshal to make sure it isn't doing anything unexpected with the current TZ.

rcowsill replied:
@thaJeztah, here's what I found: 

- The system timezone is read once per execution and cached; the new zone isn't used until the daemon restarts
- There's a quirk in the [unmarshaling of non-local times](https://play.golang.org/p/MnqaDk-7VnL) (doesn't affect checkValidParent as it uses reflect.DeepEqual)

These points don't have any negative impact on this approach or using time.Equal instead.

To clarify, I chose this approach as it only modifies behaviour in the specific case reported in #41829. Using time.Equal would fix that issue too, but it may have knock-on effects such as failing to detect a non-valid parent.

thaJeztah replied:
Thanks!

> such as failing to detect a non-valid parent.

Let me defer that one to @tonistiigi as he's more familiar with this.

We should consider consistently using UTC for these. I did a very quick check on what variants are used in the code base; looks like many paths already use `.UTC()` or `.Unix()` (which implicitly uses UTC) when reading.

I found one place that didn't use UTC; opened https://github.com/moby/moby/pull/41905 for that.

rcowsill replied:
@thaJeztah I don't know if this is important, but buildkit uses local times (time.Now()) in the history when making new images. The legacy build system uses time.Now().UTC().

thaJeztah replied:
Interesting. Wondering if there was a reason for that, or if that was overlooked 🤔. For "internal" use, it probably wouldn't make a difference if `.Equal()` is used, mostly thinking about what will end up in JSON (manifests, local storage), and if _not_ using `.UTC()` would make things less reproducible/consistent. That said; fully "reproducible" is still hard, and may require things like `SOURCE_DATE_EPOCH`

rcowsill committed:
Fix spurious error from "docker load"

"docker load" validates parent links by comparing image histories, and the
History struct has a time.Time member "Created". Time.UnmarshalJSON can read
RFC3339 timestamps with offset "+00:00", but t.MarshalJSON writes them with
offset "Z". Equivalent times in these two formats are not equal when compared
with the == operator.

This causes checkValidParent to incorrectly return false when the parent image
history contains times using offset "+00:00". In that case the history copied
to the child image will have been converted into "Z" form when marshaled out.

This patch adds an "Equal" method to History, which compares "Created" times
with t.Equal. This is used instead of reflect.DeepEqual in checkValidParent.

Signed-off-by: Rob Cowsill <42620235+rcowsill@users.noreply.github.com>

thaJeztah commented:
s390 and ppc64 failures are unrelated


@tonistiigi PTAL if this looks good to you

tonistiigi reviewed (approved):
SGTM

cpuguy83 reviewed (dismissed):
LGTM

thaJeztah reviewed (approved):
LGTM
